### PR TITLE
#2061: Updated all_gather shlo conversion into ttir and ttnn

### PR DESCRIPTION
**Pull request by tapspatel**
Captured all all_gather attributes in shlo dialect into ttir. Kept consistent naming convention for attributes found in shlo across tt-mlir stack. Updated ttir to ttnn conversion to account for dynamic cluster_axis and removed cluster axis constraint of being 1 in runtime code for all_gather.

We model all_gather in ttir dialect as the following

```
let arguments = (ins AnyRankedTensor:$input,
                         AnyRankedTensor:$output,
                         SI32Attr:$all_gather_dim,
                         OptionalAttr<SI32Attr>:$channel_handle,
                         UnitAttr:$use_global_device_ids,
                         I64ElementsAttr:$replica_groups);
```

ttnn model
```
let arguments = (ins AnyRankedTensor:$input,
                         TT_Device:$device,
                         SI32Attr:$all_gather_dim,
                         SI32Attr:$cluster_axis,
                         DefaultValuedAttr<SI32Attr, "1">:$num_links);
```

Previously, cluster_axis was set to 1 in runtime, however, if we open our mesh as the following, which is a valid test case in llmbox (ie data parallelism across 4 devices, with model parallelism across 2 devices)

```
mesh = jax.make_mesh((2, 4), ('model', 'batch'))
```

We need the ability to perform all_gather across cluster_axis=0.

GSPMD style multi-device does not propagate cluster_axis in which we perform all_gather against. It only groups devices in a replica_groups 2D tensor, which each group containing list of devices that will perform all_gather with each other. Simple algorithm was introduced to figure out cluster_axis from replica_groups.

### CODE REVIEW

**Comment by tapspatel** [comment]
I kept the default behavior for all_reduce unchanged and will follow up with changing its cluster_axis constraint from 1 to 0 or 1 in this PR: https://github.com/tenstorrent/tt-mlir/issues/2065.

**Comment by tapspatel** [comment]
Waiting on this PR to land: https://github.com/tenstorrent/tt-mlir/pull/2067 before I add silicon tests for this PR

**Comment by tapspatel** [comment]
This PR is blocked by metal mesh reshape bug (link to follow)

**Comment by tapspatel** [comment]
Currently, if we have replica_groups=1, (which doesn't mean anything computationally, ie all_gather with itself). tt-metal will throw an error. For sweep purposes, I filed an issue for them to return back the tensor unchanged: https://github.com/tenstorrent/tt-metal/issues/17804. However, in the compiler, we should optimize this at some point.

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp [hunk 1]
```diff
@@ -1348,9 +1348,73 @@ class AllGatherOpConversionPattern
                   ConversionPatternRewriter &rewriter) const override {
 
     auto device = ::ttnn::utils::getOrInsertDevice(rewriter, op);
+
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:1351]: will do, I was hoping to talk about this during our f2f, but do this is what I'll change the attributes to. We technically don't need replica_groups, channel_handle and use_global_device_ids to be modeled in TTIR right now but I kept it so we won't lose information. 

shlo
```
all_gather_dim
replica_groups
channel_id
use_global_device_ids
```

ttir
```
all_gather_dim
replica_groups
channel_handle
use_global_device_ids
**cluster_axis**
```

ttnn
```
all_gather_dim
cluster_axis
num_links
```

File: lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp [hunk 2]
```diff
@@ -341,6 +341,9 @@ class TTNNAllReduceWorkarounds : public OpRewritePattern<ttnn::AllReduceOp> {
     int32_t scatter_num = op.getScatterNum();
     Value deviceValue = op.getDevice();
     Location loc = op.getLoc();
+    uint32_t cluster_axis =
+        1; // todo: (tapspatel) hard-code to 1 to prevent any changing behaviour
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp:345]: will fix

File: lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (tapspatel) [reply lib/Conversion/TTIRToTTNN/TTIRToTTNN.cpp:1351]: fixed

File: lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (tapspatel) [reply lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp]: fixed

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 3]
```diff
@@ -1326,16 +1326,90 @@ class StableHLOToTTIRAllGatherOpConversionPattern
...
+      // the same horizontal device axis (cluster=1).
+      if (((*firstElementIt) + 1) == *secondElementIt) {
+        cluster_axis = 1;
+      } else {
+        cluster_axis = 0;
+      }
+    }
+
+    rewriter.replaceOpWithNewOp<mlir::tt::ttir::AllGatherOp>(
+        srcOp, outputType, Value(adaptor.getOperands()[0]), Value(outputTensor),
```
Comment (azecevicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1407]: You can get away without any explicit cast here (`rewriter.getSI32IntegerAttr` is needed though because it's an optional attribute).

File: lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp [hunk 4]
```diff
@@ -341,6 +341,9 @@ class TTNNAllReduceWorkarounds : public OpRewritePattern<ttnn::AllReduceOp> {
     int32_t scatter_num = op.getScatterNum();
     Value deviceValue = op.getDevice();
     Location loc = op.getLoc();
+    uint32_t cluster_axis =
```
Comment (azecevicTT) [inline lib/Dialect/TTNN/Transforms/Workarounds/TTNNWorkarounds.cpp:344]: Nit: snake_case

File: runtime/lib/ttnn/operations/ccl/all_gather.cpp [hunk 5]
```diff
@@ -13,7 +13,8 @@ namespace tt::runtime::ttnn::operations::ccl {
 void run(const ::tt::target::ttnn::AllGatherOp *op, ProgramContext &context) {
   ProgramTensorPool &tensorPool = context.getTensorPool();
   const ::ttnn::Tensor &input = tensorPool.at(op->in()->global_id());
-  int32_t gatherDim = op->dim();
+  int32_t allGatherDim = op->all_gather_dim();
+  int32_t clusterAxis = op->cluster_axis();
```
Comment (azecevicTT) [inline runtime/lib/ttnn/operations/ccl/all_gather.cpp:17]: We should respect types that got serialized, i.e. if it's signed we should store it in `int32_t`, if it's unsigned `uint32_t`.

File: test/ttmlir/Conversion/StableHLOToTTIR/ccl_ops.mlir [hunk 6]
```diff
@@ -196,8 +196,35 @@ module @all_reduce_1x32 attributes {mhlo.num_partitions = 32 : i32, mhlo.num_rep
...
+    // CHECK: "ttir.mesh_shard"
+    // CHECK-SAME: shard_dims = array<i64: -1, 1>
+    // CHECK-SAME: shard_direction = #tt.shard_direction<shard_to_full>
+    // CHECK-SAME: shard_shape = array<i64: 1, 2>
+    // CHECK-SAME: shard_type = #tt.shard_type<devices>
+    return %4 : tensor<8192x800xf32>
+  }
+  func.func private @shmap_body(%arg0: tensor<8192x400xf32>) -> (tensor<8192x400xf32> {jax.result_info = "[('batch',), ('model',)]"}) {
+    %0 = "stablehlo.all_gather"(%arg0) <{all_gather_dim = 1 : i64, channel_handle = #stablehlo.channel_handle<handle = 1, type = 1>, replica_groups = dense<[[0], [1]]> : tensor<2x1xi64>, use_global_device_ids}> : (tensor<8192x400xf32>) -> tensor<8192x400xf32>
+    // CHECK: %[[C:.*]] = "ttir.all_gather"[[C:.*]]
```
Comment (azecevicTT) [inline test/ttmlir/Conversion/StableHLOToTTIR/ccl_ops.mlir:221]: We should try to not write tests in the form of `%[[C:.*]] = "op_name"[[C:.*]]`. It's a bad practice that persisted, at least we can write just a `= "op_name"` it's more readable. See https://github.com/tenstorrent/tt-mlir/pull/1995#discussion_r1941496810. This is also relevant for all other tests.

In this case, it's important to check `cluster_index` as it's a calculated value, other values are just copied so it's not as critical to check them (though I would still do it as a form of regression test).

File: test/ttmlir/Conversion/StableHLOToTTIR/ccl_ops.mlir [hunk 7]
```diff
@@ -196,8 +196,35 @@ module @all_reduce_1x32 attributes {mhlo.num_partitions = 32 : i32, mhlo.num_rep
...
+    %3 = stablehlo.custom_call @Sharding(%2) {backend_config = "", mhlo.sharding = "{manual}"} : (tensor<8192x400xf32>) -> tensor<8192x400xf32>
+    %4 = stablehlo.custom_call @SPMDShardToFullShape(%3) {backend_config = "", mhlo.sharding = "{devices=[1,2]<=[2]}"} : (tensor<8192x400xf32>) -> tensor<8192x800xf32>
+    // CHECK: "ttir.mesh_shard"
+    // CHECK-SAME: shard_dims = array<i64: -1, 1>
+    // CHECK-SAME: shard_direction = #tt.shard_direction<shard_to_full>
+    // CHECK-SAME: shard_shape = array<i64: 1, 2>
+    // CHECK-SAME: shard_type = #tt.shard_type<devices>
+    return %4 : tensor<8192x800xf32>
+  }
+  func.func private @shmap_body(%arg0: tensor<8192x400xf32>) -> (tensor<8192x400xf32> {jax.result_info = "[('batch',), ('model',)]"}) {
```
Comment (azecevicTT) [inline test/ttmlir/Conversion/StableHLOToTTIR/ccl_ops.mlir:219]: Nit: `jax.result_info` is not needed (there are multiple places where it repeats in tests).

File: test/ttmlir/Dialect/TTNN/ccl/all_gather.mlir [hunk 8]
```diff
@@ -2,7 +2,7 @@
 module attributes {} {
   func.func @forward(%arg0: tensor<1x1x32x32xbf16>) -> tensor<1x1x32x128xbf16> {
     %0 = tensor.empty() : tensor<1x1x32x128xbf16>
-    %1 = "ttir.all_gather"(%arg0, %0) <{dim = 3 : si32}> : (tensor<1x1x32x32xbf16>, tensor<1x1x32x128xbf16>) -> tensor<1x1x32x128xbf16>
+    %1 = "ttir.all_gather"(%arg0, %0) <{all_gather_dim = 3 : si32, channel_handle = 1 : si32, replica_groups = dense<[[0, 1]]> : tensor<1x2xi64>, cluster_axis = 1 : si32}> : (tensor<1x1x32x32xbf16>, tensor<1x1x32x128xbf16>) -> tensor<1x1x32x128xbf16>
     // CHECK: %[[C:.*]] = "ttnn.all_gather"[[C:.*]]
```
Comment (azecevicTT) [inline test/ttmlir/Dialect/TTNN/ccl/all_gather.mlir:6]: Here it's less critical to make checks (but as mentioned earlier still has a great value as a regression test).

**Review by azecevicTT — COMMENTED** [review]
Few comments mostly regarding the MLIR intrinsics. Nothing critical, but tests should be covered much better.

**Review by jnie-TT — APPROVED** [review]

**Comment by tapspatel** [comment]
@azecevicTT fantastic comments! I resolved everything you mentioned, agreed with all your points. Thanks for driving better coding standards and quality. Let me know if there's any other changes you request.

**Review by sdjordjevicTT — APPROVED** [review]

File: system_desc_n300.ttsys
Comment (nsmithtt) [inline system_desc_n300.ttsys:1]: Please remove this file

**Review by nsmithtt — CHANGES_REQUESTED** [review]

File: include/ttmlir/Dialect/TTIR/IR/TTIROps.td [hunk 9]
```diff
@@ -1876,7 +1876,11 @@ def TTIR_AllGatherOp : TTIR_DPSOp<"all_gather"> {
 
     let arguments = (ins AnyRankedTensor:$input,
                          AnyRankedTensor:$output,
-                         SI32Attr:$dim);
+                         SI32Attr:$all_gather_dim,
+                         OptionalAttr<SI32Attr>:$channel_handle,
+                         UnitAttr:$use_global_device_ids,
+                         I64ElementsAttr:$replica_groups,
```
[the remark this answers is not in the document]
Comment (tapspatel) [reply include/ttmlir/Dialect/TTIR/IR/TTIROps.td:1882]: they have no use in ttmlir. I'll remove them

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 10]
```diff
@@ -1325,17 +1321,65 @@ class StableHLOToTTIRAllGatherOpConversionPattern
...
+      // Cannot have replicas of size 0, this means we are not performing the
+      // all_gather across any device.
+      return failure();
+    }
+
+    // Case where we have single devices in each replica_group (ie perform
+    // all_gather against itself which should be optimized away).
+    // We also assume we are only using our constrained mesh types (ie 1x8, 1x32
+    // etc) and cannot have (32x1, 8x1).
+    if (replicaGroupsShape[1] == 1) {
```
Comment (azecevicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1364]: I have a few questions regarding this.
1. Can we reliably rely on the assumption that we won't have shapes like `8x1` since this is still StableHLO? 
2. Do I understand this correctly that cases like `Yx1` are no-op?
3. Do we loose that information when lowered to TTIR or can we still reconstruct it?

Also nit: If you don't plan to extend these `if/else` blocks in the future, you can rewrite it as
```cpp
if (replicaGroupShape[1] != 1) {
  // firstEl
  // secondEl
 clusterAxis = (firstEl + 1) == secondEl;
}
```

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp [hunk 11]
```diff
@@ -1325,17 +1321,65 @@ class StableHLOToTTIRAllGatherOpConversionPattern
...
+      clusterAxis = 0;
+    } else {
+      auto firstElementIt = replicaGroups.begin();
+      auto secondElementIt = firstElementIt + 1;
+
+      // Case where elements in replica group differ by 1, in which they are on
+      // the same horizontal device axis (cluster=1).
+      if (((*firstElementIt) + 1) == *secondElementIt) {
+        clusterAxis = 1;
+      } else {
```
Comment (azecevicTT) [inline lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1374]: Nit: else block is not needed (even if is not needed, see previous comment).

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 3 above)
Comment (azecevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1380]: Sorry if I wasn't clear, when I said casts I meant `Value(...)`, i.e. you can call it with just `adaptor.getOperands()[0], outputTensor`, it's a more natural way to do it.

File: lib/Dialect/TTIR/IR/TTIROps.cpp [hunk 12]
```diff
@@ -1850,10 +1850,20 @@ ::mlir::LogicalResult mlir::tt::ttir::SoftmaxOp::verify() {
...
+  int32_t gatherDim = getAllGatherDim();
+  int32_t clusterAxis = getClusterAxis();
 
-  if (dim >= inputType.getRank() || dim < -inputType.getRank()) {
-    return emitOpError("Invalid dimension for all gather op.");
+  if (gatherDim >= inputType.getRank() || gatherDim < -inputType.getRank()) {
+    return emitOpError(
+        "Invalid dimension for all gather op. Gather dimension must be >= to "
+        "input tensor rank or < -input tensor rank, got gather_dim = " +
+        std::to_string(gatherDim));
```
Comment (azecevicTT) [inline lib/Dialect/TTIR/IR/TTIROps.cpp:1860]: Nit: you can use `emitOpError` as a stream, i.e. `emitOpError(...) << inputType.getRank() << "..." << gatherDim`.

File: test/ttmlir/Silicon/TTNN/llmbox/ccl/ccl_2x4.mlir [hunk 13]
```diff
@@ -2,12 +2,25 @@
 // RUN: FileCheck %s --input-file=%t.mlir
 // RUN: ttmlir-translate --ttnn-to-flatbuffer %t.mlir > %t.ttnn
 
-func.func @forward(%arg0: tensor<1x1x256x512xf32>) -> tensor<1x1x256x512xf32> {
+func.func public @all_gather_cluster0(%arg0: tensor<1x1x8192x512xf32>) -> (tensor<1x1x16384x512xf32> {jax.result_info = ""}) {
+  %0 = tensor.empty() : tensor<1x1x4096x128xf32>
+  %1 = "ttir.mesh_shard"(%arg0, %0) <{shard_dims = array<i64: 2, 3>, shard_direction = #tt.shard_direction<full_to_shard>, shard_shape = array<i64: 1, 1, 2, 4>, shard_type = #tt.shard_type<devices>}> : (tensor<1x1x8192x512xf32>, tensor<1x1x4096x128xf32>) -> tensor<1x1x4096x128xf32>
+  // CHECK: %[[C:.*]] = "ttnn.mesh_shard"[[C:.*]]
```
Comment (azecevicTT) [inline test/ttmlir/Silicon/TTNN/llmbox/ccl/ccl_2x4.mlir:8]: Remove `[[C:.*]]`. If you don't have any `CHECK` you can remove `RUN: FileCheck %s --input-file=%t.mlir`. The same goes for the rest of the tests.

**Review by azecevicTT — COMMENTED** [review]
Thanks for fixing majority of tests, there are still some left though. I'm okay with removing `FileCheck` tests from Silicon if we are just going to write them as `[[C:.*]] = op_name[[C:.*]]`.

Few nit comments, and one question regarding `replica_group` shape.

File: lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp (hunk 10 above)
Comment (tapspatel) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1364]: 1/2. right now, we have a select # of meshes we support, and any Yx1 case should be converted into 1xY. However, I believe this type of pass is complicated and needs a separate change for it. This PR addresses working configs only, while I will follow up on special case passes in another PR.

3. We loose that information due to how GSPMD multi-device is represented in compiler. Shardy (which is another multi-device compiler representation) will keep this information).

Comment (azecevicTT) [reply lib/Conversion/StableHLOToTTIR/StableHLOToTTIRPatterns.cpp:1364]: Makes sense, my concern was that we would lose the information that some op is a no-op, so we couldn't fold it in TTIR.

**Review by nsmithtt — APPROVED** [review]
